packages/doenetml-worker-rust/lib-doenetml-core/src/dep_graph/graph.rs

```diff
@@ -0,0 +1,307 @@
+///
+/// Example using `HashMap` as the `Taggable` type:
+/// ```rust
+/// # use doenetml_core::dep_graph::graph::Graph;
+/// use std::collections::HashMap;
+/// // Make a graph
+/// // a -> b -> c
+/// // a -> c
+/// let mut graph = Graph::<String, HashMap<_, _>>::new();
+/// graph.add_node("a".into());
+/// graph.add_node("b".into());
+/// graph.add_node("c".into());
+/// graph.add_edge("a".into(), "b".into());
+/// graph.add_edge("a".into(), "c".into());
+/// graph.add_edge("b".into(), "c".into());
+/// assert_eq!(graph.walk_descendants("a".into()).collect::<Vec<_>>(), vec!["b", "c"]);
+/// ```
+#[derive(Debug)]
+pub struct Graph<Node: Clone + Eq + Hash + Debug, Index: Taggable<Node, usize>> {
+    /// An `Taggable` that allows for looking up the index of a node in the graph.
+    /// I.e., used for `Node -> usize` lookups.
+    index: Index,
+    /// The nodes in the graph. Internally nodes are represented as `usize`. But `Graph` stores nodes
+    /// for the users to make its internal structure opaque.
+    nodes: Vec<Node>,
+    /// Directed edges between nodes.
+    edges: Vec<Vec<usize>>,
+    /// Reverse directed version of every edge.
+    reverse_edges: Vec<Vec<usize>>,
+}
+
+impl<Node: Clone + Eq + Hash + Debug, Index: Taggable<Node, usize>> Graph<Node, Index> {
+    /// Add a node to the graph.
+    pub fn add_node(&mut self, node: Node) {
+        let index = self.nodes.len();
+        self.index.set_tag(node.clone(), index);
+        // There should always be the same number of elements in the `nodes`, `edges`, and `reverse_edges`
+        // vectors.
+        self.nodes.push(node);
+        self.edges.push(Vec::new());
+        self.reverse_edges.push(Vec::new());
+    }
+    /// Set an edge between two nodes.
+    pub fn add_edge(&mut self, from: Node, to: Node) {
+        let &from_index = self.index.get_tag(from);
+        let &to_index = self.index.get_tag(to);
+        self.edges[from_index].push(to_index);
+        self.reverse_edges[to_index].push(from_index);
+    }
+    /// Walk through all nodes that have `node` as an ancestor. Nodes are walked in _topological_ order.
+    /// Panics if a cycle is detected.
+    pub fn walk_descendants(&self, node: Node) -> DescendantTopologicalIterator<Node> {
+        let &start_index = self.index.get_tag(node);
+        DescendantTopologicalIterator::new(&self.nodes, &self.edges, start_index)
+    }
+    /// Walk through all nodes that have `node` as a descendant. Nodes are walked in _topological_ order.
+    /// Panics if a cycle is detected.
+    pub fn walk_ancestors(&self, node: Node) -> DescendantTopologicalIterator<Node> {
+        let &start_index = self.index.get_tag(node);
+        DescendantTopologicalIterator::new(&self.nodes, &self.reverse_edges, start_index)
+    }
+
+    /// Iterate through all nodes that have `node` as an ancestor. This iterator is meant to be fast.
+    /// The order of the nodes is not guaranteed, and the same node may be yielded multiple times.
```

Comment:
Maybe include the comment that this possibility of a descendant will be yielded twice if there is a undirected cycle.